hawkular-inventory-api/src/main/java/org/hawkular/inventory/base/BasePreCommit.java

```diff
@@ -552,6 +619,30 @@ void add(EntityAndPendingNotifications<BE, ?> entity) {
             found.notifications.addAll(entity.getNotifications());
         }
 
+        ProcessingTree<BE> getChild(RelativePath childPath) {
+            Iterator<Path.Segment> segments = childPath.getPath().iterator();
+            Set<ProcessingTree<BE>> children = this.children;
+
+            CanonicalPath.Extender childCp = cp.modified();
+            ProcessingTree<BE> current = null;
+            while (segments.hasNext()) {
+                Path.Segment seg = segments.next();
+                CanonicalPath expectedCp = childCp.extend(seg.getElementType(), seg.getElementId()).get();
+                childCp = expectedCp.modified();
+
+                current = null;
+                for (ProcessingTree<BE> c : children) {
+                    if (expectedCp.equals(c.cp)) {
+                        current = c;
+                        children = c.children;
+                        break;
+                    }
+                }
+            }
+
+            return segments.hasNext() ? null : current;
```

Comment:
Good catch! it now only works by accident because usually ids on different levels are different.

The algorithm tries to go down the provided path and for each segment, it tries to find the appropriate child. If it finds the child, it compares the next segment with its children, etc.

So the missing break condition should be on current being null after the for-loop. As it is right now, it always returns the "current".